OpenZFS 8199 - multi-threaded dmu_object_alloc()

ahrens opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Motivation and Context
dmu_object_alloc() is single-threaded, so when multiple threads are
creating files in a single filesystem, they spend a lot of time waiting
for the os_obj_lock.  To improve performance of multi-threaded file
creation, we must make dmu_object_alloc() typically not grab any
filesystem-wide locks.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->


### Description
<!--- Describe your changes in detail -->

The solution is to have a “next object to allocate” for each CPU. Each
of these “next object”s is in a different block of the dnode object, so
that concurrent allocation holds dnodes in different dbufs.  When a
thread’s “next object” reaches the end of a chunk of objects (by default
4 blocks worth -- 128 dnodes), it will be reset to the per-objset
os_obj_next, which will be increased by a chunk of objects (128).  Only
when manipulating the os_obj_next will we need to grab the os_obj_lock.
This decreases lock contention dramatically, because each thread only
needs to grab the os_obj_lock briefly, once per 128 allocations.

Work sponsored by Intel Corp.

Closes #4703

### How Has This Been Tested?
This results in a 70% performance improvement to multi-threaded object
creation (where each thread is creating objects in its own directory),
from 67,000/sec to 115,000/sec, with 8 CPUs.

![image](https://cloud.githubusercontent.com/assets/799124/25919741/7db70bf4-3584-11e7-8cd0-82019d7b027e.png)

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

behlendorf commented on module/zfs/dmu_object.c:
```diff
@@ -50,6 +59,9 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
 	boolean_t restarted = B_FALSE;
+	uint64_t *cpuobj = &os->os_obj_next_array[CPU_SEQID %
+	    os->os_obj_next_array_len];
+	int dnodes_per_chunk = 1 << dmu_object_alloc_chunk_shift;
```
I'm not exactly sure what the situation is on Illumos.  But under Linux `CPU_SEQID % os->os_obj_next_array_len]` won't guarantee that there aren't multiple CPUs concurrently in this function operating on the same `cpuobj`.  I'm assuming the same it true for Illumos because between the atomic updates of `cpuobj` and the taking of `dn->dn_struct_rwlock` lock this looks to be handled nicely.

For the record this could happen as result on CPU hot plugging when additional processors come online and `os->os_obj_next_array` isn't resized.  Or more likely the process is moved to a different CPU during `schedule()`.   Or potentially it's rescheduled at any time when using a real-time pre-emptible kernel build.

ahrens replied:
Correct, the cpuobj should typically typically different on different CPUs, but if it happens to be the same that is handled.

behlendorf commented on module/zfs/dmu_object.c:
```diff
@@ -341,4 +384,10 @@ EXPORT_SYMBOL(dmu_object_free);
+/* BEGIN CSTYLED */
+module_param(dmu_object_alloc_chunk_shift, int, 0644);
+MODULE_PARM_DESC(dmu_object_alloc_chunk_shift,
+	"CPU-specific allocator grabs 2^N objects at once");
```
Are you sure it's safe to allow this to be changed at run time?  This would result in a miscalculation of where the per-cpu blocks of objects ends.  Although it looks like that would result in some wasted called to `dnode_hold_impl()` and then it would sort itself out.  Right?

ahrens replied:
It would go back to the global allocator at the "wrong" time, potentially resulting in multiple allocators hitting the same block concurrently, but this situation should rectify itself after one iteration.  But if you're concerned, is there a way to make it only settable at boot time?

behlendorf replied:
The `0644` in the `module_param()` call is for the mode bits of the entry in `/sys/module/zfs/parameters/`.  Changing it to `0444` would make it only settable at module load time.  But in general it's nice to have everything easily modifiable at runtime so I'm happy leaving it as is.  I just wanted to double check my understanding this was OK.

ahrens commented on module/zfs/dmu_object.c:
```diff
@@ -32,6 +32,15 @@
+ * determined to be the lowest value that eliminates the measurable effect
+ * of lock contention from this code path.
+ */
+int dmu_object_alloc_chunk_shift = 7;
```
The tunable is expressed in the number of objects (technically number of slots).

I assume you mean that I should say "dnode slots" rather than "dnodes".

behlendorf replied:
"dnode slots" or "object slots" would be clearer to me than "objects".  I also just noticed a typo in this comment,  `s/which s 4/which is 4/`.

"The default is to grab 128 object slots, which is 4 blocks worth."

ahrens commented:
Thanks @behlendorf.  I updated the comment as suggested.

ahrens commented on module/zfs/dmu_object.c:
```diff
@@ -58,83 +70,114 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
+			 * from the beginning at most once per txg. If we
+			 * still can't allocate from that L1 block, search
+			 * for an empty L0 block, which will quickly skip
+			 * to the end of the metadnode if the no nearby L0
```
sure thing.

ahrens commented on module/zfs/dmu_object.c:
```diff
@@ -58,83 +70,114 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
+	if (dnodes_per_chunk > L1_dnode_count)
+		dnodes_per_chunk = L1_dnode_count;
+
+	object = atomic_inc_64_nv(cpuobj);
```
Hmm.  It seems like it would be better if *cpuobj was the next object to try to allocate, rather than the last object that was allocated.  Since as you said we could (in theory) be allocating a different number of slots than the previous allocation.  I'll see about changing the code to do that.

nedbass replied:
I thought about that. That's the way it previously worked with `os->os_object_next`. My concern was that if we don't atomically update `*cpuobj` here then it becomes possible for concurrent threads to enter the for loop having the same value in `object`. But I guess we are still protected in that case by taking `dn_struct_rwlock` below.

ahrens replied:
I think we're still OK, and since it's rare that multiple threads would concurrently use the same *cpuobj, the performance impact of having to go to the next object should be OK.  I've made the change, can you take a look at the new code?

nedbass commented on module/zfs/dmu_object.c:
```diff
@@ -58,83 +70,114 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
+				if (error == 0) {
+					object = offset >> DNODE_SHIFT;
+					os->os_obj_next =
+					    object + dnodes_per_chunk;
```
Should we update `*cpuobj` here as well?

ahrens replied:
I don't think we *need* to, since we'll update it later when we decide which object to actually allocate.  Worst case, if we change CPU's, another thread could also see the old *cpuobj and then redo this work (dnode_next_offset()).  I guess for that reason maybe we should update it?  It's not very expensive to do so.

nedbass replied:
Well I asked because it's updated above after we update `os->os_obj_next`, so it made me wonder why not here too? It's more of a code clarity issue for me than a performance issue. Maybe move the update of `*cpuobj` to just before we drop `os->os_obj_lock` below to cover both cases?

ahrens replied:
that makes sense, I'll do it.

nedbass commented on module/zfs/dmu_object.c:
```diff
@@ -58,83 +70,114 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
+			mutex_enter(&os->os_obj_lock);
+			ASSERT0(P2PHASE(os->os_obj_next, dnodes_per_chunk));
+			object = os->os_obj_next;
+			os->os_obj_next += dnodes_per_chunk;
```
It's a little confusing to have an `os->os_obj_next` and an `os->os_obj_next_array`. Perhaps we should rename the former to `os->obj_next_chunk`.

ahrens replied:
Sure.  Maybe `os_obj_next_chunk` and `os_obj_next_percpu`?

nedbass replied:
Yes, I find that easier to understand.

ahrens commented:
@nedbass I think I've addressed your feedback, could you take one last look?  Also could you let me know if the failed tests are something I need to look at?  I clicked around a bit and didn't find any errors that I could understand :-)

nedbass commented:
@ahrens, thanks for addressing the feedback. We had also proposed renaming `os->os_obj_next` to `os->os_obj_next_chunk`. If you still agree with that naming choice then that is still unaddressed.

Regarding the failures, it looks like we tripped the assert `ASSERT0(P2PHASE(os->os_obj_next, dnodes_per_chunk));`

```
[ 2328.090441] SPL: using hostid 0x00000000
[ 2328.134694] VERIFY3(0 == ((os->os_obj_next) & ((dnodes_per_chunk) - 1))) failed (0 == 2)
[ 2328.141374] PANIC at dmu_object.c:93:dmu_object_alloc_dnsize()
[ 2328.146082] Showing stack for process 30735
[ 2328.150109] CPU: 0 PID: 30735 Comm: zpool Tainted: P           OE  ------------   3.10.0-514.16.1.el7.x86_64 #1
[ 2328.157723] Hardware name: Xen HVM domU, BIOS 4.2.amazon 11/11/2016
[ 2328.161625]  ffffffffa07b5104 000000006cb6831b ffff8801bb66b8f8 ffffffff81686ac3
[ 2328.166631]  ffff8801bb66b908 ffffffffa04b2e34 ffff8801bb66ba90 ffffffffa04b2eff
[ 2328.171484]  000000006cb6831b ffff880100000030 ffff8801bb66baa0 ffff8801bb66ba40
[ 2328.176748] Call Trace:
[ 2328.179740]  [<ffffffff81686ac3>] dump_stack+0x19/0x1b
[ 2328.185527]  [<ffffffffa04b2e34>] spl_dumpstack+0x44/0x50 [spl]
[ 2328.191503]  [<ffffffffa04b2eff>] spl_panic+0xbf/0xf0 [spl]
[ 2328.196223]  [<ffffffff81698eb1>] ? ftrace_call+0x5/0x2f
[ 2328.200263]  [<ffffffff8168e305>] ? _raw_spin_unlock+0x5/0x30
[ 2328.206245]  [<ffffffffa0621b78>] ? dbuf_rele_and_unlock+0x1f8/0x5c0 [zfs]
[ 2328.211998]  [<ffffffffa063437f>] dmu_object_alloc_dnsize+0x2af/0x470 [zfs]
[ 2328.217338]  [<ffffffffa06c954b>] zap_create_norm_dnsize+0x2b/0x60 [zfs]
[ 2328.225800]  [<ffffffffa06c95f2>] zap_create+0x22/0x30 [zfs]
[ 2328.230656]  [<ffffffffa06fb1fa>] zfs_create_fs+0x84a/0xa90 [zfs]
[ 2328.235431]  [<ffffffff81698eb1>] ? ftrace_call+0x5/0x2f
[ 2328.240017]  [<ffffffff81698eb1>] ? ftrace_call+0x5/0x2f
[ 2328.244874]  [<ffffffffa0668a33>] dsl_pool_create+0x1e3/0x500 [zfs]
[ 2328.250504]  [<ffffffffa0692aef>] spa_create+0x5af/0xb30 [zfs]
[ 2328.254597]  [<ffffffffa06e0862>] zfs_ioc_pool_create+0x152/0x270 [zfs]
[ 2328.258937]  [<ffffffffa06e0715>] ? zfs_ioc_pool_create+0x5/0x270 [zfs]
[ 2328.262938]  [<ffffffffa06e16a7>] zfsdev_ioctl+0x537/0x660 [zfs]
[ 2328.267654]  [<ffffffffa06e1170>] ? pool_status_check+0xf0/0xf0 [zfs]
[ 2328.272678]  [<ffffffff81212555>] do_vfs_ioctl+0x2d5/0x4b0
[ 2328.276149]  [<ffffffff812127d1>] SyS_ioctl+0xa1/0xc0
[ 2328.279415]  [<ffffffff81697189>] system_call_fastpath+0x16/0x1b
[ 2520.231124] INFO: task zpool:30735 blocked for more than 120 seconds.
```

ahrens commented:
@nedbass Sounds good I'll change it to `os_obj_next_chunk`.

Ah, I must have missed this link, I assume that's where you found the stack trace:
http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20Mainline%20%28TEST%29/builds/2504/steps/shell_6/logs/console
I'll investigate that.

nedbass commented:
@ahrens Yes, check the console log anytime you see tests fails due to timeout, as that is where you'll find any failed ASSERTs and kernel stack traces.

ahrens commented:
@nedbass I think there are two ways that this assertion can fail:

The `restarted` logic could make us find the next block of dnodes, rather than the next L1.  So the alignment of `os_obj_next` could be only a single block’s worth (32) — which is less than `dmu_object_alloc_chunk_shift` (which is between a block’s worth an a L1's worth).

`zfs_create_fs()` modifies `os_obj_next`, setting it to exactly `2`.  Looks like that change was also part of the large blocks project.

Any advice on how to handle either of these?

nedbass commented:
> The restarted logic could make us find the next block of dnodes, rather than the next L1. So the alignment of os_obj_next could be only a single block’s worth (32) — which is less than dmu_object_alloc_chunk_shift (which is between a block’s worth an a L1's worth).

My thought here is to just mask off the lower bits when assigning `os_obj_next_chunk` so it's always a multiple of `dnodes_per_chunk`.

> zfs_create_fs() modifies os_obj_next, setting it to exactly 2. Looks like that change was also part of the large dnode project.

I think we can remove the code that sets it to 2. That was needed when we were advancing `os_obj_next` to the next dnode block in case `dmu_object_next()` returned an error, which it did when `object == 0`  at fs creation time. But with your re-working of the code that case now looks to be handled by the global allocator logic.

The former consequence of not initializing `os_obj_next` to 2 was that object numbers of files in a brand-new dataset would begin at 32. If you observe new object numbers beginning < 32 (typically 7) after removing that code then it should be fine to leave it out.

This does bring up the point that there is no handling of errors from `dmu_object_next()`. What happens if it returns ESRCH?  It looks like that might result in an infinite loop. Do we think that's not possible when searching for a hole? Perhaps that should be documented in the form of an ASSERT or a comment.

ahrens commented:
@nedbass 
> My thought here is to just mask off the lower bits when assigning os_obj_next_chunk so it's always a multiple of dnodes_per_chunk.

Will do.

> I think we can remove the code that sets it to 2. 

Will do.

> This does bring up the point that there is no handling of errors from dmu_object_next().

I like your logic of advancing to the next L0.  That way at least it can be sure to make progress, and eventually rescan if necessary.  I'll put that back in place.

nedbass commented:
> You mentioned that dmu_object_next() returned an error when object == 0. I didn't see where in the code that would happen, so I'm having trouble understanding why it seems to be getting an error on the MOS but not on regular objects.

@ahrens, I bet it's hitting this case in `dnode_next_offset()`:

```
        if (dn->dn_phys->dn_nlevels == 0) {
                error = SET_ERROR(ESRCH);
                goto out;
        }
```

If I recall, that's why I made the comment (removed in your last commit) about the metadnode's dnode_phys_t structure not being initialized yet. Not sure what the best way to deal with that is, which is why I ended up advancing `os_obj_next` to 2. That way the initial `dnode_hold_impl()` call would succeed and thereby avoid the call to `dmu_object_next()` during dataset creation.

ahrens commented:
@nedbass I see, the problem isn't calling dmu_object_next(object=0), but rather calling dmu_object_next() at all when the objset has not been synced to disk yet.  And this would happen when we first call dmu_object_alloc() because the call to dnode_hold_impl(object=0) would fail.

I'm still trying to figure out why this doesn't happen for ordinary filesystems, for the objects allocated from zfs_create_fs().  But I don't think it's a big deal either way.

behlendorf commented:
@ahrens when you're happy with this go ahead and squash the commits and add your signed-off-by.  We'll queue this up for some additional local testing too.

nedbass commented:
@ahrens  also don't forget about renaming `os_obj_next` to `os_obj_next_chunk` when you rebase :smile:

tonyhutter commented:
@ahrens  I'm seeing ANSI codes in your commit message:
```
    The solution is to have a <E2><80><9C>next object to allocate<E2><80><9D> for each CPU. Each
    of these <E2><80><9C>next object<E2><80><9D>s is in a different block of the dnode object, so
    that concurrent allocation holds dnodes in different dbufs.  When a
    thread<E2><80><99>s <E2><80><9C>next object<E2><80><9D> reaches the end of a chunk of objects (by default
    4 blocks worth -- 128 dnodes), it will be reset to the per-objset
    os_obj_next, which will be increased by a chunk of objects (128).  
   ...
```

tonyhutter commented:
I benchmarked this using the same test setup as https://github.com/zfsonlinux/zfs/pull/5935#issuecomment-306263532 (16 core Xeon, ramdisk pool, results averaged over 3 runs).  The numbers look great!

File creates per second:

| Threads | Before | After | Speedup |
| --- | --- | --- | --- |
| 1 | 17928.2 | 17796.7 | -0.73% |
| 2 | 29141.5 | 31927.1 | 9.56% |
| 4 | 54211.7 | 56868.7 | 4.90% |
| 8 | 85954.5 | 99901.9 | 16.23% |
| 16 | 108344.7 | 150460.7 | 38.87% |
| 32 | 112364.9 | 154242.2 | 37.27% |
| 64 | 111778.3 | 156167.0 | 39.71% |
| 128 | 113122.6 | 158316.7 | 39.95% |
| 256 | 113899.4 | 154484.2 | 35.63% |

tonyhutter commented:
More results, this time using real hardware.  This is a SSD pool used by one of our lustre MDS nodes (16 core):
```
 	jet1        ONLINE       0     0     0
	  mirror-0  ONLINE       0     0     0
	    U0      ONLINE       0     0     0
	    U1      ONLINE       0     0     0
	    L2      ONLINE       0     0     0
	  mirror-1  ONLINE       0     0     0
	    L0      ONLINE       0     0     0
	    L1      ONLINE       0     0     0
	    U2      ONLINE       0     0     0
```
File creates/sec:

| Threads | Before | After | Speedup |
| --- | --- | --- | --- |
| 1 | 10255 | 10469 | 2% |
| 2 | 18168 | 19707 | 8% |
| 4 | 32316 | 35476 | 10% |
| 8 | 52143 | 56630 | 9% |
| 16 | 57917 | 76697 | 32% |
| 32 | 58791 | 76357 | 30% |
| 64 | 58643 | 77782 | 33% |
| 128 | 59471 | 79658 | 34% |
| 256 | 59713 | 82078 | 37% |
